_posts/2019-12-20-differential-privacy/differential_privacy.Rmd

```diff
@@ -0,0 +1,470 @@
+
+```{r}
+# number of timesteps used in the second dimension
+num_timesteps <- 12
+
+# number of observations to be used for the training set
+# a round number for easier checking!
+train_max <- 4000
+
+# also keep track of actual number of training and test observations
+n_train <- train_max - num_timesteps + 1
+n_test <- n_total - train_max - num_timesteps + 1
+```
+
+Here, then, are the basic building blocks that will go into the final training and test datasets.
+
+```{r}
+acc_train <-
+  average_and_make_sequences(acc[1:(train_max * 64), ], 64, num_timesteps)
+bvp_train <-
+  average_and_make_sequences(bvp[1:(train_max * 128), , drop = FALSE], 128, num_timesteps)
+eda_train <-
+  average_and_make_sequences(eda[1:(train_max * 8), , drop = FALSE], 8, num_timesteps)
+temp_train <-
+  average_and_make_sequences(temp[1:(train_max * 8), , drop = FALSE], 8, num_timesteps)
+
+
+acc_test <-
+  average_and_make_sequences(acc[(train_max * 64 + 1):nrow(acc), ], 64, num_timesteps)
+bvp_test <-
+  average_and_make_sequences(bvp[(train_max * 128 + 1):nrow(bvp), , drop = FALSE], 128, num_timesteps)
+eda_test <-
+  average_and_make_sequences(eda[(train_max * 8 + 1):nrow(eda), , drop = FALSE], 8, num_timesteps)
+temp_test <-
+  average_and_make_sequences(temp[(train_max * 8 + 1):nrow(temp), , drop = FALSE], 8, num_timesteps)
+```
+
+Now put all predictors together:
+
+```{r}
+# all predictors
+x_train <- zip_datasets(acc_train, bvp_train, eda_train, temp_train) %>%
+  dataset_map(function(...)
+    tf$concat(list(...), axis = 1L))
+
+x_test <- zip_datasets(acc_test, bvp_test, eda_test, temp_test) %>%
+  dataset_map(function(...)
+    tf$concat(list(...), axis = 1L))
+```
+
+On the ground truth side, as alluded to before, we leave out the first eleven values in each case: 
+
+```{r}
+y_train <- tensor_slices_dataset(label[num_timesteps:train_max] %>% tf$cast(tf$float32))
+
+y_test <- tensor_slices_dataset(label[(train_max + num_timesteps):nrow(label)] %>% tf$cast(tf$float32))
+
+```
+
+Zip predictors and targets together, configure shuffling/batching, and the datasets are complete:
+
+```{r}
+ds_train <- zip_datasets(x_train, y_train)
+ds_test <- zip_datasets(x_test, y_test)
+
+batch_size <- 32
+
+ds_train <- ds_train %>% 
+  dataset_shuffle(n_train) %>%
+  # dataset_repeat is needed because of pre-TF 2 style
+  # hopefully at a later time, the code can run eagerly and this is no longer needed
+  dataset_repeat() %>%
+  dataset_batch(batch_size, drop_remainder = TRUE)
+
+ds_test <- ds_test %>%
+  # see above reg. dataset_repeat
+  dataset_repeat() %>%
+  dataset_batch(batch_size)
+
+
+```
+
+With data manipulations as complicated as the above, it's always worthwhile checking some pipeline outputs. We can do that using the usual `reticulate::as_iterator` magic, provided that for this test run, we _don't_ disable V2 behavior. (Just restart the R session between a "pipeline checking" and the later modeling runs.)
```

Comment:
```suggestion
With data manipulations as complicated as the above, it's always worthwhile to check some pipeline outputs. We can do that using the usual `reticulate::as_iterator` magic, provided that for this test run, we _don't_ disable TF V2 behavior. (Just restart the R session between a "pipeline checking" and the later modeling runs.)
```

Reply:
Added the TF, since there's plenty the reader's holding in their brains between the previous discussion of it and now!